File: image_filter.cpp
/*
 * Ejercicio 1: Aplicar un filtro a múltiples imágenes
 * Implementación paralela usando OpenMP para procesar múltiples imágenes
 */

#include <iostream>
#include <vector>
#include <string>
#include <chrono>
#include <fstream>
#include <random>
#include <omp.h>
#include <opencv2/opencv.hpp>

using namespace std;
using namespace cv;

// Estructura para almacenar información de una imagen
struct ImageData {
    string filename;
    Mat image;
    double processing_time;
};

// Función para aplicar filtro de escala de grises a una imagen
void applyGrayscaleFilter(ImageData& img_data) {
    auto start_time = chrono::high_resolution_clock::now();
    
    // Convertir a escala de grises
    Mat grayscale;
    cvtColor(img_data.image, grayscale, COLOR_BGR2GRAY);
    
    // Guardar imagen procesada
    string output_filename = "output_images/grayscale_" + img_data.filename;
    imwrite(output_filename, grayscale);
    
    auto end_time = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::microseconds>(end_time - start_time);
    img_data.processing_time = duration.count() / 1000000.0; // Convertir a segundos
}

// Versión secuencial del filtro de imágenes
vector<ImageData> applyGrayscaleFilterSequential(const vector<ImageData>& images) {
    vector<ImageData> results = images;
    
    for (auto& img : results) {
        applyGrayscaleFilter(img);
    }
    
    return results;
}

// Versión paralela del filtro de imágenes usando OpenMP
vector<ImageData> applyGrayscaleFilterParallel(const vector<ImageData>& images) {
    vector<ImageData> results = images;
    
    #pragma omp parallel for
    for (int i = 0; i < static_cast<int>(results.size()); ++i) {
        applyGrayscaleFilter(results[i]);
    }
    
    return results;
}

// Función para crear imágenes de muestra
vector<ImageData> createSampleImages(int num_images = 8, int width = 200, int height = 200) {
    vector<ImageData> images;
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> dis(0, 255);
    
    // Crear directorio de imágenes de muestra
    system("mkdir -p sample_images");
    
    for (int i = 0; i < num_images; ++i) {
        // Crear imagen con colores aleatorios
        Mat image(height, width, CV_8UC3);
        
        for (int y = 0; y < height; ++y) {
            for (int x = 0; x < width; ++x) {
                image.at<Vec3b>(y, x) = Vec3b(dis(gen), dis(gen), dis(gen));
            }
        }
        
        string filename = "sample_image_" + to_string(i + 1) + ".png";
        string filepath = "sample_images/" + filename;
        imwrite(filepath, image);
        
        ImageData img_data;
        img_data.filename = filename;
        img_data.image = image.clone();
        images.push_back(img_data);
    }
    
    return images;
}

// Función principal de benchmark
void benchmarkImageFiltering() {
    cout << "=== EJERCICIO 1: FILTRO DE IMÁGENES (C++ con OpenMP) ===" << endl << endl;
    
    // Crear imágenes de muestra
    cout << "Creando imágenes de muestra..." << endl;
    vector<ImageData> images = createSampleImages(8);
    cout << "Se crearon " << images.size() << " imágenes de muestra" << endl << endl;
    
    // Crear directorio de salida
    system("mkdir -p output_images");
    
    // Benchmark versión secuencial
    cout << "Ejecutando versión secuencial..." << endl;
    auto start_time = chrono::high_resolution_clock::now();
    vector<ImageData> sequential_results = applyGrayscaleFilterSequential(images);
    auto end_time = chrono::high_resolution_clock::now();
    auto sequential_duration = chrono::duration_cast<chrono::microseconds>(end_time - start_time);
    double sequential_time = sequential_duration.count() / 1000000.0;
    
    cout << "Tiempo secuencial: " << fixed << setprecision(4) << sequential_time << " segundos" << endl;
    cout << "Tiempo promedio por imagen: " << sequential_time / images.size() << " segundos" << endl;
    
    // Limpiar archivos de salida para la prueba paralela
    system("rm -f output_images/*");
    
    // Benchmark versión paralela
    cout << endl << "Ejecutando versión paralela..." << endl;
    start_time = chrono::high_resolution_clock::now();
    vector<ImageData> parallel_results = applyGrayscaleFilterParallel(images);
    end_time = chrono::high_resolution_clock::now();
    auto parallel_duration = chrono::duration_cast<chrono::microseconds>(end_time - start_time);
    double parallel_time = parallel_duration.count() / 1000000.0;
    
    cout << "Tiempo paralelo: " << parallel_time << " segundos" << endl;
    cout << "Tiempo promedio por imagen: " << parallel_time / images.size() << " segundos" << endl;
    
    // Calcular speedup
    double speedup = (parallel_time > 0) ? sequential_time / parallel_time : 0;
    cout << endl << "Speedup: " << speedup << "x" << endl;
    
    int num_threads = omp_get_max_threads();
    double efficiency = (speedup / num_threads) * 100;
    cout << "Eficiencia: " << efficiency << "%" << endl;
    cout << "Número de threads utilizados: " << num_threads << endl;
    
    // Mostrar resultados detallados
    cout << endl << "Resultados detallados:" << endl;
    for (const auto& result : parallel_results) {
        cout << "  " << result.filename << ": " << result.processing_time << "s" << endl;
    }
    
    cout << endl << "Imágenes procesadas exitosamente: " << parallel_results.size() << endl;
    cout << "Archivos guardados en: output_images/" << endl << endl;
}

int main() {
    // Configurar OpenMP
    omp_set_num_threads(omp_get_max_threads());
    
    cout << "OpenMP disponible: " << omp_get_max_threads() << " threads" << endl;
    cout << "Compilador OpenMP: " << _OPENMP << endl << endl;
    
    benchmarkImageFiltering();
    
    return 0;
}

File: image_filter.py
"""
Ejercicio 1: Aplicar un filtro a múltiples imágenes
Implementación paralela usando concurrent.futures para procesar múltiples imágenes
"""

import os
import time
from concurrent.futures import ProcessPoolExecutor, ThreadPoolExecutor
from PIL import Image
import numpy as np
from typing import List, Tuple


def apply_grayscale_filter(image_path: str, output_path: str) -> Tuple[str, float]:
    """
    Aplica un filtro de escala de grises a una imagen
    
    Args:
        image_path: Ruta de la imagen de entrada
        output_path: Ruta donde guardar la imagen procesada
    
    Returns:
        Tupla con el nombre del archivo y tiempo de procesamiento
    """
    start_time = time.time()
    
    try:
        # Cargar imagen
        image = Image.open(image_path)
        
        # Convertir a escala de grises
        grayscale_image = image.convert('L')
        
        # Guardar imagen procesada
        grayscale_image.save(output_path)
        
        processing_time = time.time() - start_time
        return os.path.basename(image_path), processing_time
        
    except Exception as e:
        print(f"Error procesando {image_path}: {e}")
        return os.path.basename(image_path), 0.0


def apply_grayscale_filter_sequential(image_paths: List[str], output_dir: str) -> List[Tuple[str, float]]:
    """
    Versión secuencial del filtro de imágenes
    
    Args:
        image_paths: Lista de rutas de imágenes
        output_dir: Directorio de salida
    
    Returns:
        Lista de tuplas con resultados
    """
    results = []
    
    for image_path in image_paths:
        filename = os.path.basename(image_path)
        output_path = os.path.join(output_dir, f"grayscale_{filename}")
        result = apply_grayscale_filter(image_path, output_path)
        results.append(result)
    
    return results


def apply_grayscale_filter_parallel(image_paths: List[str], output_dir: str, max_workers: int = None) -> List[Tuple[str, float]]:
    """
    Versión paralela del filtro de imágenes usando ProcessPoolExecutor
    
    Args:
        image_paths: Lista de rutas de imágenes
        output_dir: Directorio de salida
        max_workers: Número máximo de workers (None = automático)
    
    Returns:
        Lista de tuplas con resultados
    """
    # Preparar rutas de salida
    output_paths = []
    for image_path in image_paths:
        filename = os.path.basename(image_path)
        output_path = os.path.join(output_dir, f"grayscale_{filename}")
        output_paths.append(output_path)
    
    # Procesar en paralelo
    with ProcessPoolExecutor(max_workers=max_workers) as executor:
        results = list(executor.map(apply_grayscale_filter, image_paths, output_paths))
    
    return results


def create_sample_images(num_images: int = 5, size: Tuple[int, int] = (200, 200)) -> List[str]:
    """
    Crea imágenes de muestra para testing
    
    Args:
        num_images: Número de imágenes a crear
        size: Tamaño de las imágenes
    
    Returns:
        Lista de rutas de las imágenes creadas
    """
    os.makedirs("sample_images", exist_ok=True)
    image_paths = []
    
    for i in range(num_images):
        # Crear imagen con colores aleatorios
        image_array = np.random.randint(0, 256, (size[1], size[0], 3), dtype=np.uint8)
        image = Image.fromarray(image_array)
        
        filename = f"sample_image_{i+1}.png"
        image_path = os.path.join("sample_images", filename)
        image.save(image_path)
        image_paths.append(image_path)
    
    return image_paths


def benchmark_image_filtering():
    """
    Ejecuta benchmark comparando versión secuencial vs paralela
    """
    print("=== EJERCICIO 1: FILTRO DE IMÁGENES ===\n")
    
    # Crear imágenes de muestra
    print("Creando imágenes de muestra...")
    image_paths = create_sample_images(8)
    print(f"Se crearon {len(image_paths)} imágenes de muestra\n")
    
    # Crear directorio de salida
    os.makedirs("output_images", exist_ok=True)
    
    # Benchmark versión secuencial
    print("Ejecutando versión secuencial...")
    start_time = time.time()
    sequential_results = apply_grayscale_filter_sequential(image_paths, "output_images")
    sequential_time = time.time() - start_time
    
    print(f"Tiempo secuencial: {sequential_time:.4f} segundos")
    print(f"Tiempo promedio por imagen: {sequential_time/len(image_paths):.4f} segundos")
    
    # Limpiar archivos de salida para la prueba paralela
    for _, _ in sequential_results:
        pass  # Los archivos ya están creados
    
    # Benchmark versión paralela
    print("\nEjecutando versión paralela...")
    start_time = time.time()
    parallel_results = apply_grayscale_filter_parallel(image_paths, "output_images")
    parallel_time = time.time() - start_time
    
    print(f"Tiempo paralelo: {parallel_time:.4f} segundos")
    print(f"Tiempo promedio por imagen: {parallel_time/len(image_paths):.4f} segundos")
    
    # Calcular speedup
    speedup = sequential_time / parallel_time if parallel_time > 0 else 0
    print(f"\nSpeedup: {speedup:.2f}x")
    print(f"Eficiencia: {(speedup / os.cpu_count()) * 100:.1f}%")
    
    # Mostrar resultados detallados
    print("\nResultados detallados:")
    for filename, proc_time in parallel_results:
        print(f"  {filename}: {proc_time:.4f}s")
    
    print(f"\nImágenes procesadas exitosamente: {len(parallel_results)}")
    print("Archivos guardados en: output_images/\n")


if __name__ == "__main__":
    benchmark_image_filtering()
